testng-eclipse-plugin/src/main/org/testng/eclipse/launch/TestNGLaunchConfigurationDelegate.java

```diff
@@ -391,35 +390,25 @@ private Version getRuntimeTestNGVersion(List<String> classpath)
       File f = new File(cp);
       // only check jar file name starts with 'testng'
       if (f.isFile() && f.getName().startsWith("testng")) {
-        Version ver = null;
         try (JarFile jarFile = new JarFile(f)) {
           Manifest mf = jarFile.getManifest();
           Attributes mainAttrs = mf.getMainAttributes();
-          String sn = mainAttrs.getValue("Bundle-SymbolicName");
-          if ("org.testng".equals(sn)) {
-            ver = new Version(mainAttrs.getValue("Bundle-Version"));
+          if ("org.testng".equals(mainAttrs.getValue("Bundle-SymbolicName"))) {
+            return new Version(mainAttrs.getValue("Bundle-Version"));
           }
 
-          if (ver == null) {
-            String mc = mainAttrs.getValue("Main-Class");
-            if ("org.testng.TestNG".equals(mc)) {
-              ver = new Version(mainAttrs.getValue("Implementation-Version"));
-            }
+          if ("org.testng.TestNG".equals(mainAttrs.getValue("Main-Class"))) {
+            return new Version(mainAttrs.getValue("Implementation-Version"));
           }
 
-          if (ver == null) {
-            String st = mainAttrs.getValue("Specification-Title");
-            if ("testng".equals(st)) {
-              ver = new Version(mainAttrs.getValue("Specification-Version"));
-            }
+          if ("testng".equals(mainAttrs.getValue("Specification-Title"))) {
+            return new Version(mainAttrs.getValue("Specification-Version"));
           }
         } catch (Exception e) {
           TestNGPlugin.log(e);
         }
 
-        if (ver != null) {
-          return ver;
-        }
+        TestNGPlugin.log(cp + " was the potential target, but no canonical version found in MANIFEST.MF");

```

Comment:
if the jar file name starts with 'testng', it's pretty much likely be the testng.jar, if we can't parse the version from its MANIFEST.MF, we log the file name for further diagnose.